src/operators.jl

```diff
@@ -74,16 +83,56 @@ function Op(f, T=Any; iscommutative=false)
         error("User-defined reduction operators are not supported on 32-bit Windows.\nSee https://github.com/JuliaParallel/MPI.jl/issues/246 for more details.")
     end
     w = OpWrapper{typeof(f),T}(f)
-    fptr = @cfunction($w, Cvoid, (Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cint}, Ptr{MPI_Datatype}))
-    
-    op = Op(OP_NULL.val, fptr)
+    cfunc = @cfunction($w, Cvoid, (Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cint}, Ptr{MPI_Datatype}))
+
+    op = Op(OP_NULL.val, cfunc)
     # int MPI_Op_create(MPI_User_function* user_fn, int commute, MPI_Op* op)
     @mpichk ccall((:MPI_Op_create, libmpi), Cint,
                   (Ptr{Cvoid}, Cint, Ptr{MPI_Op}),
-                  fptr, iscommutative, op)
+                  cfunc, iscommutative, op)
 
     refcount_inc()
     finalizer(free, op)
     return op
 end
 
+"""
+    @declareOp(op, T[, iscommutative])
+
+Declare a Julia function `op` to be used as a custom MPI operator [`Op`](@ref) for
+variables of type `T`. This will create the [`Op`](@ref) object and define an appropriate
+constructor method to `Op`. The `iscommutative` argument indicates to MPI whether or not
+MPI can assume the operation is commutative (default is `false`).
+
+The usage of this macro is optional: the main advantage of this is that will avoid the use
+of a closure (see ["Closure cfunctions" in the Julia
+manual](https://docs.julialang.org/en/v1/manual/calling-c-and-fortran-code/#Closure-cfunctions-1),
+which may offer some performance advantages.
+
+This should only be called once per combination of `op` and `T`, and should be at the
+top-level (e.g. not inside a function). It can be safely used before `MPI.Init()` and
+inside a precompiled module.
+"""
+macro declareOp(f, T, iscommutative=false)
+    opwrap = gensym(:opwrap) # we need to manually gensym for use with `@cfunction` macro
+    quote
+        if !Base.issingletontype(typeof($(esc(f))))
+            error("@declareOp macro can only be used with instances of singleton types")
+        end        
+        const op = Op(OP_NULL.val, nothing)
+        const $(esc(opwrap)) = OpWrapper{typeof($(esc(f))),$(esc(T))}($(esc(f)))
+        function initop()
+            fptr = @cfunction($opwrap, Cvoid, (Ptr{Cvoid}, Ptr{Cvoid}, Ptr{Cint}, Ptr{MPI_Datatype}))
```

Comment:
Yes that should work with PPC. It is the dynamic version of `@cfunction` that requires an LLVM feature that doesn't work on PPC.